### PR TITLE
Brian/rebalancing token misc

### PR DESCRIPTION
**Pull request by bweick**

### CODE REVIEW

File: contracts/core/RebalancingSetToken.sol [hunk 1]
```diff
@@ -171,21 +171,22 @@ contract RebalancingSetToken is
         require(msg.sender == manager);
 
         // New proposal cannot be made during a rebalance period
-        require(rebalanceState == State.Default);
+        require(rebalanceState != State.Rebalance);
```
Comment (asoong) [inline contracts/core/RebalancingSetToken.sol]: Is this so they can re-propose?

File: contracts/core/RebalancingSetToken.sol [hunk 2]
```diff
@@ -171,21 +171,22 @@ contract RebalancingSetToken is
...
 
         // Make sure enough time has passed from last rebalance to start a new proposal
         require(block.timestamp >= lastRebalanceTimestamp.add(rebalanceInterval));
 
+        // Check that new proposed Set is valid Set created by Core
+        address core = ISetFactory(factory).core();
```
Comment (asoong) [inline contracts/core/RebalancingSetToken.sol]: `core` doesn't look re-used. Insert `ISetFactory(factory).core()` directly into the next expression

File: contracts/core/RebalancingSetToken.sol [hunk 3]
```diff
@@ -327,9 +331,11 @@ contract RebalancingSetToken is
     function getComponents()
         external
         view
-        returns(address[1])
+        returns(address[])
     {
-        return [currentSet];
+        address[] memory components = new address[](1);
```
Comment (asoong) [inline contracts/core/RebalancingSetToken.sol]: this seems verbose. with only one value that is already known, you can just initialize the array with it I think: https://ethereum.stackexchange.com/questions/28047/solidity-initialize-memory-array-of-fixed-size

File: contracts/core/RebalancingSetToken.sol [hunk 4]
```diff
@@ -340,9 +346,11 @@ contract RebalancingSetToken is
     function getUnits()
         external
         view
-        returns(uint256[1])
+        returns(uint256[])
     {
-        return [unitShares];
+        uint256[] memory units = new uint256[](1);
```
Comment (asoong) [inline contracts/core/RebalancingSetToken.sol]: same

File: test/core/extensions/coreIssuance.spec.ts [hunk 5]
```diff
@@ -18,7 +21,15 @@ import {
 import { ether } from '../../../utils/units';
 import { IssuanceComponentDeposited } from '../../../utils/contract_logs/core';
 import { assertTokenBalance, expectRevertError } from '../../../utils/tokenAssertions';
-import { DEFAULT_GAS, DEPLOYED_TOKEN_QUANTITY, ZERO } from '../../../utils/constants';
+import {
+  DEFAULT_GAS,
+  DEPLOYED_TOKEN_QUANTITY,
+  NULL_ADDRESS,
```
Comment (asoong) [inline test/core/extensions/coreIssuance.spec.ts]: How is this working? I think you need to rebase, `NULL_ADDRESS` is part of set-protocol-utils now

File: test/core/extensions/coreIssuance.spec.ts [hunk 6]
```diff
@@ -316,7 +330,242 @@ contract('CoreIssuance', accounts => {
...
+        setTokenFactory.address,
+        componentAddresses,
+        componentUnits,
+        naturalUnit,
+      );
+
+      initialShareRatio = DEFAULT_UNIT_SHARES;
+      rebalancingNaturalUnit = DEFAULT_REBALANCING_NATURAL_UNIT;
+      const initialSet = setToken.address;
+      const manager = managerAccount;
```
Comment (asoong) [inline test/core/extensions/coreIssuance.spec.ts]: Can you move this into a utils method or something? Something like:
`generateRebalancingSetCreatorCallData` and you can move all of these one time use constants into there.

File: test/core/extensions/coreIssuance.spec.ts [hunk 7]
```diff
@@ -316,7 +330,242 @@ contract('CoreIssuance', accounts => {
...
+          { from: ownerAccount, gas: DEFAULT_GAS },
+        );
+      });
+
+      it('should revert', async () => {
+        await expectRevertError(subject());
+      });
+    });
+
+    // describe('when the quantity is not a multiple of the natural unit of the set', async () => {
```
Comment (asoong) [inline test/core/extensions/coreIssuance.spec.ts]: ?

File: test/core/extensions/coreIssuance.spec.ts [hunk 8]
```diff
@@ -442,6 +691,148 @@ contract('CoreIssuance', accounts => {
...
+      componentUnits = _.map(components, () => ether(4)); // Multiple of naturalUnit
+      setToken = await coreWrapper.createSetTokenAsync(
+        core,
+        setTokenFactory.address,
+        componentAddresses,
+        componentUnits,
+        naturalUnit,
+      );
+
+      initialShareRatio = DEFAULT_UNIT_SHARES;
```
Comment (asoong) [inline test/core/extensions/coreIssuance.spec.ts]: same as above for this setup

File: test/core/extensions/coreIssuance.spec.ts [hunk 9]
```diff
@@ -316,7 +330,242 @@ contract('CoreIssuance', accounts => {
...
+      const initialSet = setToken.address;
+      const manager = managerAccount;
+      const proposalPeriod = ONE_DAY_IN_SECONDS;
+      const rebalanceInterval = ONE_DAY_IN_SECONDS;
+      const callData = Utils.bufferArrayToHex([
+        Utils.paddedBufferForPrimitive(manager),
+        Utils.paddedBufferForBigNumber(proposalPeriod),
+        Utils.paddedBufferForBigNumber(rebalanceInterval),
+      ]);
+      rebalancingToken = await coreWrapper.createRebalancingTokenAsync(
```
Comment (asoong) [inline test/core/extensions/coreIssuance.spec.ts]: for this `createRebalancingTokenAsync` method, make some of the parameters default to these constants so you dont need to keep defining them above

File: test/core/extensions/coreIssuance.spec.ts [hunk 10]
```diff
@@ -442,6 +691,148 @@ contract('CoreIssuance', accounts => {
...
+      beforeEach(async () => {
+        subjectQuantityToRedeem = ether(3);
+      });
+
+      it('should revert', async () => {
+        await expectRevertError(subject());
+      });
+    });
+
+    // describe('when the quantity is not a multiple of the natural unit of the set', async () => {
```
Comment (asoong) [inline test/core/extensions/coreIssuance.spec.ts]: ?

File: test/core/setToken.spec.ts [hunk 11]
```diff
@@ -390,145 +390,4 @@ contract('SetToken', accounts => {
       });
     });
   });
-
-  describe('#transfer', async () => {
```
Comment (asoong) [inline test/core/setToken.spec.ts]: why are you removing these?

File: contracts/core/RebalancingSetToken.sol (hunk 1 above)
Comment (bweick) [reply contracts/core/RebalancingSetToken.sol]: Si papi

File: contracts/core/RebalancingSetToken.sol (hunk 3 above)
Comment (asoong) [reply contracts/core/RebalancingSetToken.sol]: try `return new address[currentSet];`

File: contracts/core/RebalancingSetToken.sol (hunk 2 above)
Comment (bweick) [reply contracts/core/RebalancingSetToken.sol]: Yeah felt like that would be super messy though. Preferred the clarity.

File: test/core/setToken.spec.ts (hunk 11 above)
Comment (bweick) [reply test/core/setToken.spec.ts]: It's unnecessary...it basically calls the inherited xfer fxn except has a check to make sure that the recipient address isn't the smart contract itself. Just no need to have it in my mind.

File: contracts/core/RebalancingSetToken.sol (hunk 1 above)
Comment (asoong) [reply contracts/core/RebalancingSetToken.sol]: 👍

File: contracts/core/RebalancingSetToken.sol (hunk 2 above)
Comment (asoong) [reply contracts/core/RebalancingSetToken.sol]: going for gas optimization at the contract level, extra memory consumed here

File: contracts/core/RebalancingSetToken.sol (hunk 3 above)
Comment (bweick) [reply contracts/core/RebalancingSetToken.sol]: The problem is that, that doesn't return a dynamic typed array which wouldn't conform to the getComponents and getUnits interface

**Pull request merged**